### PR TITLE
Disabled Resource warning for explicit detach of grpc task

### PR DESCRIPTION
**Pull request by MounikaBattu17**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nidaqmx-python/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?

- Added a condition in `__del__` method in `task.py` to not to throw Resource warning for grpc task.
- Updated `test_resource_warnings.py` to test the no warning case for grpc.

### Why should this Pull Request be merged?
This PR fixes [Bug 2383312](https://dev.azure.com/ni/DevCentral/_workitems/edit/2383312): nidaqmx-python reports ResourceWarning when detaching from a newly created gRPC task

### What testing has been done?
Ran `poetry run pytest` for `test_resource_warnings.py`

### CODE REVIEW

**Review by keer1111 — APPROVED** [review]

**Review by zhindes — APPROVED** [review]

File: tests/legacy/test_resource_warnings.py [hunk 1]
```diff
@@ -17,10 +19,21 @@ def test_task_double_close(self, init_kwargs):
...
+
+    @pytest.mark.grpc_only
+    def test_explicit_detach_grpc_task(self, init_kwargs):
+        """Test to validate grpc leak tasks."""
+        task = nidaqmx.Task(**init_kwargs)
+
+        with warnings.catch_warnings(record=True) as warnings_raised:
+            task.__del__()
+            if warnings_raised:
+                pytest.fail()
```
Comment (bkeryan) [inline tests/legacy/test_resource_warnings.py:39]: Split the assertion into a separate section (arrange/act/assert sections are separated by a blank line) and use the `assert` statement. The `warnings_raised` variable retains its value after exiting the `with` block.

```
        task = nidaqmx.Task(**init_kwargs)

        with warnings.catch_warnings(record=True) as warnings_raised:
            task.__del__()

        assert len(warnings_raised) == 0
```

File: tests/legacy/test_resource_warnings.py [hunk 2]
```diff
@@ -17,10 +19,21 @@ def test_task_double_close(self, init_kwargs):
             with nidaqmx.Task(**init_kwargs) as task:
                 task.close()
 
+    @pytest.mark.library_only
     def test_unclosed_task(self, init_kwargs):
```
Comment (bkeryan) [inline tests/legacy/test_resource_warnings.py:23]: Hmm, these tests actually leak tasks. Perhaps they should save the interpreter and task handle and call clear_task at the end.
image

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: tests/legacy/test_resource_warnings.py [hunk 3]
```diff
@@ -17,10 +19,21 @@ def test_task_double_close(self, init_kwargs):
...
         """Test to validate unclosed tasks."""
         task = nidaqmx.Task(**init_kwargs)
         # Since __del__ is not guaranteed to be called, for the purposes of
         # consistent test results call __del__ manually.
         with pytest.warns(DaqResourceWarning):
             task.__del__()
```
Comment (bkeryan) [inline tests/legacy/test_resource_warnings.py:29]: BTW, pytest reports a warning from this test:

```
================================================== warnings summary ===================================================
tests/legacy/test_resource_warnings.py::TestResourceWarnings::test_unclosed_task[grpc_init_kwargs]
  D:\dev\nidaqmx-python\generated\nidaqmx\task.py:93: ResourceWarning: Task of name "_unnamedTask<0>" was not explicitly closed before it was destructed. Resources on the task device may still be reserved.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
==================================== 2 passed, 745 deselected, 1 warning in 2.46s =====================================
```

I don't think `task.__del__()` actually destroys the task object. It's just a callback that the GC invokes before collecting the object.

File: tests/legacy/test_resource_warnings.py [hunk 4]
```diff
@@ -17,10 +19,21 @@ def test_task_double_close(self, init_kwargs):
...
+
+    @pytest.mark.grpc_only
+    def test_explicit_detach_grpc_task(self, init_kwargs):
```
[the remark this answers is not in the document]
Comment (MounikaBattu17) [reply tests/legacy/test_resource_warnings.py]: Moved these two test methods to new `test_task_resource_warnings.py` file in `tests/component` folder.

[the remark this answers is not in the document]
Comment (MounikaBattu17) [reply tests/legacy/test_resource_warnings.py]: Followed the naming conventions in `test_task_resource_warnigs.py`

File: tests/legacy/test_resource_warnings.py (hunk 1 above)
Comment (MounikaBattu17) [reply tests/legacy/test_resource_warnings.py]: Updated.

File: tests/legacy/test_resource_warnings.py (hunk 2 above)
Comment (MounikaBattu17) [reply tests/legacy/test_resource_warnings.py]: Closed all the leaking tasks.

File: tests/legacy/test_resource_warnings.py (hunk 3 above)
Comment (MounikaBattu17) [reply tests/legacy/test_resource_warnings.py]: I updated the tests but there is one discrepancy found, I performed task leaking on both library task and grpc task, expected library interpreter task warns but grpc task doesn't warn. Tests passed as expected. There is one warning by GC from library interpreter task, but in test summary, the warning is categorized in grpc task test method
```
import warnings
import pytest

import nidaqmx
from nidaqmx import DaqResourceWarning

@pytest.mark.library_only
def test__library_task__leak_task__raised_resource_warning(init_kwargs):
    task = nidaqmx.Task("LibraryTask", **init_kwargs)

    with pytest.warns(DaqResourceWarning):
        task.__del__()
    
    # Closing the task resolves the warning.
    # task.close()

@pytest.mark.grpc_only
def test__grpc_task__leak_task__not_raised_resource_warning(init_kwargs):
    task = nidaqmx.Task("GrpcTask", **init_kwargs)

    with warnings.catch_warnings(record=True) as warnings_raised:
        task.__del__()

    assert len(warnings_raised) == 0
```

```
====================================================== warnings summary ====================================================== 
tests/component/test_sample_task_res_warns.py::test__grpc_task__leak_task__not_raised_resource_warning[grpc_init_kwargs]       
  C:\Users\battu.mounika\source\repos\TAF\NI-DAQmxPython New\nidaqmx-python\generated\nidaqmx\task.py:94: ResourceWarning: Task of name "LibraryTask" was not explicitly closed before it was destructed. Resources on the task device may still be reserved. 
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================ 2 passed, 1 warning in 3.28s ================================================
```

Closing the Library interpreter task resolves the warning. But not sure why the library interpreter warning is caught in grpc task test method. Please share your thoughts?

Comment (MounikaBattu17) [reply tests/legacy/test_resource_warnings.py]: I added a Test case `test__closed_task__del_task__resource_warning_not_raised` in `test_task_resource_warnings.py` to validate `__del__` of a closed task.

File: tests/component/test_task_resource_warnings.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
+import warnings
+import nidaqmx
+
+import pytest
+
+from nidaqmx.errors import DaqResourceWarning
+
+
+@pytest.mark.library_only
+def test__unclosed_task__leak_task__resource_warning_raised(task):
```
Comment (bkeryan) [inline tests/component/test_task_resource_warnings.py:10]: In test function names, separate given/when/then with 3 underscores.

File: tests/component/test_task_resource_warnings.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    """Test to validate unclosed tasks."""
```
Comment (bkeryan) [inline tests/component/test_task_resource_warnings.py:11]: Don't add docstrings to test functions that use given/when/then naming.

File: tests/component/test_task_resource_warnings.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    interpreter = task._interpreter
+    task_handle = task._handle
```
Comment (bkeryan) [inline tests/component/test_task_resource_warnings.py:13]: Sorry, I suggested this before I realized that we can still close the task after calling `task.__del__()`.

The `task` fixture owns the task and automatically closes it, so delete the interpreter/task_handle stuff. Calling `interpreter.clear_task()` followed by `task.clear()` is a double clear.

For the 1st two test functions that use the `task` fixture, all you need is the "act" section (`catch_warnings` and `__del__`). The 3rd test function that uses `init_kwargs` is fine.

File: tests/unit/test_task.py [hunk 8]
```diff
@@ -153,5 +154,23 @@ def test___close_on_exit___leak_task___raises_resource_warning(interpreter: Mock
...
+    grpc_options = create_grpc_options(mocker)
+
+    task = Task("GrpcTask", grpc_options=grpc_options)
+
+    with warnings.catch_warnings(record=True) as warnings_raised:
+        task.__del__()
+    
+    assert len(warnings_raised) == 0
+
+    task.close()
```
Comment (bkeryan) [inline tests/unit/test_task.py:176]: The test function should have a max of 3 sections (arrange, act, assert), separated by blank lines. This one has 5.

```suggestion
    expect_create_task(interpreter, "GrpcTaskHandle")
    expect_get_task_name(interpreter, "GrpcTask")
    grpc_options = create_grpc_options(mocker)
    task = Task("GrpcTask", grpc_options=grpc_options)

    with warnings.catch_warnings(record=True) as warnings_raised:
        task.__del__()
    
    assert len(warnings_raised) == 0
    task.close()
```

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: tests/legacy/test_resource_warnings.py (hunk 3 above)
Comment (bkeryan) [reply tests/legacy/test_resource_warnings.py]: If you call `task.__del__()` and then leak the task, the GC calls `task.__del__()` again outside the `with pytest.warns()` block.

`test__closed_task__del_task__resource_warning_not_raised` looks fine.

File: tests/legacy/test_resource_warnings.py (hunk 2 above)
Comment (bkeryan) [reply tests/legacy/test_resource_warnings.py]: Oops, I should have updated/closed this when I realized we can still call `task.close()` after `task.__del__()`.

File: tests/legacy/test_resource_warnings.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (bkeryan) [reply tests/legacy/test_resource_warnings.py]: Ok but I raised a new issue about the number of underscores (I know, it's subtle and I got it wrong in the past)

File: tests/unit/test_task.py (hunk 8 above)
Comment (MounikaBattu17) [reply tests/unit/test_task.py]: Applied the suggestion.

File: tests/component/test_task_resource_warnings.py (hunk 5 above)
Comment (MounikaBattu17) [reply tests/component/test_task_resource_warnings.py]: Added underscore in function names.

**Pull request merged**